### DISCUSSION TITLE
Passing parameters to Jobs with version 2.2.400.0

### DISCUSSION THREAD

**Discussion by AlvarMayor** [quartznet/quartznet#502] [General]
### Version: 2.2.400.0

I am new to Quartz, and I "inherited" an old project, without any documentation, written by an unknows developer and I need to expand it, adding a new property to the Parameters class passed to the job.
In this project a job is created like below:

```
SyncUrlsJobParameters ParamEsiti = new SyncUrlsJobParameters();
ParamEsiti.IDCategoria = 1; //New field
serializer = new JavaScriptSerializer();
value = serializer.Serialize(ParamEsiti);
JobDataMap syncUrlsJobDataEsiti = new JobDataMap();
syncUrlsJobDataEsiti.Add("Parameters", Encoding.UTF8.GetBytes(value));
IJobDetail syncUrlsJobEsiti = JobBuilder.Create<SyncUrlsJob>()
                 .WithIdentity("SyncUrlsJob")
                 .WithDescription("Sincronizzazione fonti WEB ESITI")
                 .UsingJobData(syncUrlsJobDataEsiti)
                 .StoreDurably(true)                 
                 .Build();
```

app.config is like below:
```
<quartz>
    <add key="quartz.scheduler.instanceName" value="EwinJobScheduler" />
    <add key="quartz.threadPool.threadCount" value="10" />
    <add key="quartz.threadPool.threadPriority" value="2" />
    <add key="quartz.jobStore.misfireThreshold" value="60000" />
    <add key="quartz.jobStore.type" value="Quartz.Impl.AdoJobStore.JobStoreTX, Quartz" />
    <add key="quartz.jobStore.driverDelegateType" value="Quartz.Impl.AdoJobStore.StdAdoDelegate, Quartz" />
    <add key="quartz.jobStore.tablePrefix" value="QRTZ_" />
    <add key="quartz.jobStore.dataSource" value="myDS" />
    <add key="quartz.dataSource.myDS.connectionString" value="Data Source=SQL2012;Initial Catalog=myDB;User ID=sa;password=mypwd" />
    <add key="quartz.dataSource.myDS.provider" value="SqlServer-20" />
    <add key="quartz.jobStore.useProperties" value="false" />
  </quartz>
```

In this project i have an "ancestor" generic class, and all real Jobs inherits from this class:
```
public abstract class JobBase<T> : IJob, IDisposable where T : JobParameters, new()
    {
        public abstract void Run();
        
        public void Execute(IJobExecutionContext context)
        {           
            var jobDataMap = context.MergedJobDataMap;   
            //This call loads "something" from the JOB_DATA field of table QRTZ_TRIGGERS table
        }
    }
```

When execution reach this point in jobDataMap the parameters class is the OLD class, without the new field, exactly the one loaded fron the trigger JOB_DATA field.

Any attempt to pass parameters to the job fails: at this point in jobDataMap my parameters are always lost.

First question:
How can I pass parameters (in ANY way) to a Job and retrieve them in the Run method?
Second question:
Or, how can I change the JOB_DATA field in QRTZ_TRIGGERS table?

**Comment by lahma** [comment]
JSON serialization should be quite resilient to adding new fields, so maybe write some tests against data written to file and then deserialize etc. Of course the new values won't be present for the old data but deserialization should work fine.

I also suggest reading the documentation about [job parameters](https://www.quartz-scheduler.net/documentation/quartz-2.x/tutorial/more-about-jobs.html).

**Comment by AlvarMayor** [comment]
Yes, serialization works, but i need the new parameters.
How can I change the JOB_DATA field in QRTZ_TRIGGERS table adding the new parameters?

**Comment by lahma** [comment]
You need to load the triggers/jobs, update the contained deserialized object to have value in the new field and then update the jobs/triggers.

**Comment by AlvarMayor** [comment]
Ok, I understand the way, can you give me a piece of code as sample?
